### PR TITLE
LDAP password authentication

### PR DESCRIPTION
**Pull request by liggitt**
- [x] Add LDAP password authenticator
- [x] audit godeps
- [x] Godoc
- [x] Unit tests
- [x] Integration tests
- [x] create story for QE - https://trello.com/c/UC2rGgJ2/
- [ ] open ansible issue
- [ ] open doc issue

Sample configuration to test TLS/StartTLS (`ldap.ca.crt` is the Red Hat CA - https://password.corp.redhat.com/newca.crt):
```
  identityProviders:
  - challenge: true
    login: true
    name: "secureldap"
    provider:
      apiVersion: v1
      attributes:
        email:
        - mail
        id:
        - dn
        name:
        - cn
        preferredUsername:
        - uid
      bindDN: ""
      bindPassword: ""
      ca: ldap.ca.crt
      kind: LDAPPasswordIdentityProvider
      insecure: false
      url: "ldap://ldap.rdu.redhat.com/ou=users,dc=redhat,dc=com?uid"
```

Sample configuration to test `insecure: true` (login with "newton" / "password"):
```
  identityProviders:
  - challenge: true
    login: true
    name: "insecureldap"
    provider:
      apiVersion: v1
      attributes:
        id:
        - dn
        name:
        - cn
        preferredUsername:
        - uid
      bindDN: "cn=read-only-admin,dc=example,dc=com"
      bindPassword: "password"
      ca: ""
      kind: LDAPPasswordIdentityProvider
      insecure: true
      url: "ldap://ldap.forumsys.com/dc=example,dc=com?uid"
```

### CODE REVIEW

**Comment by liggitt** [comment]
@deads2k still working on an integration test, but code is ready for review

File: pkg/auth/authenticator/password/ldappassword/url.go [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+	"net/url"
+	"strings"
+
+	"github.com/go-ldap/ldap"
+)
+
+// LDAPURL holds a parsed RFC 2255 URL
+type LDAPURL struct {
+	// Scheme is ldap or ldaps
+	Scheme string
```
Comment (deads2k) [inline pkg/auth/authenticator/password/ldappassword/url.go]: type for this with enumerated values.

File: pkg/auth/authenticator/password/ldappassword/url.go [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+	// HostPort is the host:port of the LDAP server
+	HostPort string
```
Comment (deads2k) [inline pkg/auth/authenticator/password/ldappassword/url.go]: is it valuable to keep these together?

File: pkg/auth/authenticator/password/ldappassword/url.go [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+	// The DN of the branch of the directory where all searches should start from
+	BaseDN string
+	// The attribute to search for
+	QueryAttribute string
+	// The scope of the search. Can be either 'one', 'sub', or 'base'
+	Scope int
```
Comment (deads2k) [inline pkg/auth/authenticator/password/ldappassword/url.go]: godoc wrong?  I like the idea of an enumerated string.  I haven't read to see why this is an `int`.

File: pkg/auth/authenticator/password/ldappassword/url.go (hunk 2 above)
Comment (liggitt) [inline pkg/auth/authenticator/password/ldappassword/url.go]: I was sort of mirroring URL... they get parsed together in `url.Parse`, they get used together in `Dial`...

File: pkg/auth/authenticator/password/ldappassword/url.go (hunk 1 above)
Comment (liggitt) [inline pkg/auth/authenticator/password/ldappassword/url.go]: again, mirroring `url.URL`

File: pkg/auth/authenticator/password/ldappassword/url.go (hunk 3 above)
Comment (liggitt) [inline pkg/auth/authenticator/password/ldappassword/url.go]: yeah, wrong godoc... it's using ints from the ldap library (which are named, but not typed, LDAP constants)

File: pkg/auth/authenticator/password/ldappassword/url.go [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+	}
+
+	// Set host (default to localhost to match mod_auth_ldap)
+	opts.HostPort = parsedURL.Host
+	if len(opts.HostPort) == 0 {
+		opts.HostPort = "localhost"
+	}
+
+	// Add port if needed
+	if _, _, err := net.SplitHostPort(opts.HostPort); err != nil {
```
Comment (deads2k) [inline pkg/auth/authenticator/password/ldappassword/url.go]: Since the very next thing you do is split the HostPort, why bother keeping them together at all?

File: pkg/auth/authenticator/password/ldappassword/url.go [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+	case ProtocolLDAP, ProtocolLDAPS:
+		// ok
+	default:
+		return LDAPURL{}, fmt.Errorf("invalid scheme %q", parsedURL.Scheme)
+	}
+
+	// Set host (default to localhost to match mod_auth_ldap)
+	opts.HostPort = parsedURL.Host
+	if len(opts.HostPort) == 0 {
+		opts.HostPort = "localhost"
```
Comment (deads2k) [inline pkg/auth/authenticator/password/ldappassword/url.go]: This seems weird.  We're unwilling to default scheme to `ldaps`, but we will default `HostPort`?  Actually, why do we default this?

File: pkg/auth/authenticator/password/ldappassword/url.go [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+	}
+
+	// Set base dn (default to "")
+	// url.Parse() already percent-decodes the path
+	opts.BaseDN = strings.TrimLeft(parsedURL.Path, "/")
+
+	// Split query
+	var attributes, scope, filter, extensions string
+	parts := strings.Split(parsedURL.RawQuery, "?")
+	if len(parts) > 0 {
```
Comment (deads2k) [inline pkg/auth/authenticator/password/ldappassword/url.go]:
```
switch len(parts){
case 0:
case 3:
    //get filter
    fallthrough
case 2:
   // get scope
   fallthrough
//etc
default:
    // too many options
}
```

File: pkg/auth/authenticator/password/ldappassword/url.go [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+		default:
+			return LDAPURL{}, fmt.Errorf("no default port for scheme %q", opts.Scheme)
+		}
+	}
+
+	// Set base dn (default to "")
+	// url.Parse() already percent-decodes the path
+	opts.BaseDN = strings.TrimLeft(parsedURL.Path, "/")
+
+	// Split query
```
Comment (deads2k) [inline pkg/auth/authenticator/password/ldappassword/url.go]: include a full example for us poor saps who don't remember it all offhand?

File: pkg/auth/authenticator/password/ldappassword/url.go (hunk 1 above)
Comment (deads2k) [inline pkg/auth/authenticator/password/ldappassword/url.go]: > again, mirroring url.URL

`URL` is generic for scheme.  This is not.

File: pkg/auth/authenticator/password/ldappassword/url.go (hunk 3 above)
Comment (deads2k) [inline pkg/auth/authenticator/password/ldappassword/url.go]: > yeah, wrong godoc... it's using ints from the ldap library (which are named, but not typed, LDAP constants)

I'd like named ints to make this idiot proof.  The guys dealing with the ldap libraries live and breath it.  People like me hit it at the edges.  Keep me on the straight, narrow, and idiot-proof

File: pkg/auth/authenticator/password/ldappassword/url.go [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+	}
+	if len(parts) > 3 {
+		extensions = parts[3]
+	}
+	if len(parts) > 4 {
+		return LDAPURL{}, fmt.Errorf("too many query options %q", parsedURL.RawQuery)
+	}
+
+	// Set query attribute to first attribute (default to uid to match mod_auth_ldap)
+	opts.QueryAttribute = strings.Split(attributes, ",")[0]
```
Comment (deads2k) [inline pkg/auth/authenticator/password/ldappassword/url.go]: if `len(parts) == 0`, this panics, right?

File: pkg/auth/authenticator/password/ldappassword/url.go (hunk 3 above)
Comment (liggitt) [inline pkg/auth/authenticator/password/ldappassword/url.go]: I never reference magic numbers... this is where it's assigned:
```
	switch scope {
	case "", "sub":
		opts.Scope = ldap.ScopeWholeSubtree
	case "one":
		opts.Scope = ldap.ScopeSingleLevel
	case "base":
		opts.Scope = ldap.ScopeBaseObject
	default:
		return LDAPURL{}, fmt.Errorf("invalid scope %q", scope)
	}
```

File: pkg/auth/authenticator/password/ldappassword/url.go (hunk 4 above)
Comment (liggitt) [inline pkg/auth/authenticator/password/ldappassword/url.go]: renamed to `Host` to match URL, the split is only there to see if it contains a port

File: pkg/auth/authenticator/password/ldappassword/url.go (hunk 5 above)
Comment (liggitt) [inline pkg/auth/authenticator/password/ldappassword/url.go]: > why do we default this?

matching `mod_auth_ldap`. You can do `ldap:///o=foo,o=bar...`

File: pkg/auth/authenticator/password/ldappassword/url.go [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+		opts.Filter = "(objectClass=*)"
+	}
+	if _, err := ldap.CompileFilter(opts.Filter); err != nil {
+		return LDAPURL{}, fmt.Errorf("invalid filter: %v", err)
+	}
+
+	// Fail if there are any critical extensions, since we don't support any
+	if len(extensions) > 0 {
+		for _, extension := range strings.Split(extensions, ",") {
+			exttype := strings.SplitN(extension, "=", 2)[0]
```
Comment (deads2k) [inline pkg/auth/authenticator/password/ldappassword/url.go]: if the extension is malformed, this code might panic.

File: pkg/auth/authenticator/password/ldappassword/url.go (hunk 3 above)
Comment (liggitt) [inline pkg/auth/authenticator/password/ldappassword/url.go]: updated godoc with named constants

File: pkg/auth/authenticator/password/ldappassword/url.go [hunk 10]
```diff
@@ -0,0 +1,146 @@
...
+	// Scheme is ldap or ldaps
+	Scheme Scheme
+	// Host is the host:port of the LDAP server
+	Host string
+	// The DN of the branch of the directory where all searches should start from
+	BaseDN string
+	// The attribute to search for
+	QueryAttribute string
+	// The scope of the search. Can be ldap.ScopeWholeSubtree, ldap.ScopeSingleLevel, or ldap.ScopeBaseObject
+	Scope int
```
Comment (deads2k) [inline pkg/auth/authenticator/password/ldappassword/url.go]: If this were private or in a private struct, I'd have no issue.  Having a non-enumerated enum in a public field of a public type worries me.  Maybe I haven't spent enough time in Go to just be happy....

Comment (liggitt) [inline pkg/auth/authenticator/password/ldappassword/url.go]: typified, just for you

File: pkg/auth/authenticator/password/ldappassword/url.go (hunk 9 above)
Comment (liggitt) [inline pkg/auth/authenticator/password/ldappassword/url.go]: how so? SplitN returns a slice of length 1..N

File: pkg/auth/authenticator/password/ldappassword/url.go (hunk 8 above)
Comment (liggitt) [inline pkg/auth/authenticator/password/ldappassword/url.go]: no, dividing the string "" by "," results in [""]

File: pkg/auth/authenticator/password/ldappassword/url.go (hunk 1 above)
Comment (liggitt) [inline pkg/auth/authenticator/password/ldappassword/url.go]: typified

File: pkg/auth/authenticator/password/ldappassword/url.go (hunk 2 above)
Comment (liggitt) [inline pkg/auth/authenticator/password/ldappassword/url.go]: renamed to `Host`

File: pkg/auth/authenticator/password/ldappassword/ldap.go [hunk 11]
```diff
@@ -0,0 +1,227 @@
...
+	default:
+		err = fmt.Errorf("unsupported scheme %q", a.options.URL.Scheme)
+	}
+	if err != nil {
+		return nil, false, err
+	}
+	defer l.Close()
+
+	// Upgrade to TLS if needed
+	if !secured && !a.options.Insecure {
```
Comment (deads2k) [inline pkg/auth/authenticator/password/ldappassword/ldap.go]: move this into the switch and eliminate `secure`?

File: pkg/auth/authenticator/password/ldappassword/ldap.go [hunk 12]
```diff
@@ -0,0 +1,227 @@
...
+
+// Authenticator validates username/passwords against an LDAP v3 server
+type Authenticator struct {
+	providerName string
+	options      Options
+	mapper       authapi.UserIdentityMapper
+}
+
+// New returns an authenticator which will validate usernames/passwords using LDAP.
+func New(providerName string, options Options, mapper authapi.UserIdentityMapper) (authenticator.Password, error) {
```
Comment (deads2k) [inline pkg/auth/authenticator/password/ldappassword/ldap.go]: These options can be invalid for our purposes, I think we should have a validate method.  Particularly for `ldapurl.schema`, `insecure`, `tlsconfig`

File: pkg/auth/authenticator/password/ldappassword/ldap.go (hunk 11 above)
Comment (liggitt) [inline pkg/auth/authenticator/password/ldappassword/ldap.go]: refactored to a `connect` function... the only tricky bit is if the TLS upgrade fails, we have to close the connection ourselves

File: pkg/auth/authenticator/password/ldappassword/url.go (hunk 6 above)
Comment (liggitt) [inline pkg/auth/authenticator/password/ldappassword/url.go]: fallthrough with switch in reverse length order looks uglier to me, but updated

File: pkg/auth/authenticator/password/ldappassword/url.go (hunk 7 above)
Comment (liggitt) [inline pkg/auth/authenticator/password/ldappassword/url.go]: added

File: pkg/auth/authenticator/password/ldappassword/ldap.go (hunk 12 above)
Comment (liggitt) [inline pkg/auth/authenticator/password/ldappassword/ldap.go]: I'm not inclined to double validate here

**Comment by liggitt** [comment]
comments addressed

File: pkg/auth/authenticator/password/ldappassword/url.go (hunk 8 above)
Comment (liggitt) [inline pkg/auth/authenticator/password/ldappassword/url.go]: also, the unit tests exercise the empty cases quite thoroughly

File: pkg/auth/authenticator/password/ldappassword/ldap.go (hunk 12 above)
Comment (deads2k) [inline pkg/auth/authenticator/password/ldappassword/ldap.go]: > I'm not inclined to double validate here

I'm missing the first validation from https://github.com/openshift/origin/pull/3558/files#diff-fada3c2d88a9a92dde56e2d9c58e75a1R467.  You're saying validated via the particular callpath in this location based on the master config options?

File: pkg/auth/authenticator/password/ldappassword/ldap.go [hunk 13]
```diff
@@ -0,0 +1,240 @@
...
+
+	// If specified, bind the username/password for search phase
+	if len(a.options.BindDN) > 0 {
+		if err := l.Bind(a.options.BindDN, a.options.BindPassword); err != nil {
+			return nil, false, err
+		}
+	}
+
+	// & together the filter specified in the LDAP options with the user-specific filter
+	filter := fmt.Sprintf("(&%s(%s=%s))",
```
Comment (deads2k) [inline pkg/auth/authenticator/password/ldappassword/ldap.go]: Not that I think any sane person would do it, but do we want to support arbitrary filters that allow for multiple ands/ors/nots, etc?  Things like, "I want to use this ldap, but only people belonging to this department have access and I can control that with a clever filter in my authenticator."

File: pkg/auth/authenticator/password/ldappassword/ldap.go [hunk 14]
```diff
@@ -0,0 +1,240 @@
...
+	attrs.Insert(a.options.AttributePreferredUsername...)
+	attrs.Insert(a.options.AttributeID...)
+
+	// Search for LDAP record
+	searchRequest := ldap.NewSearchRequest(
+		a.options.URL.BaseDN,     // base dn
+		int(a.options.URL.Scope), // scope
+		ldap.NeverDerefAliases,   // deref
+		2,            // size limit, we want to know if this is not unique, but don't want the entire tree
+		0,            // time limit
```
Comment (deads2k) [inline pkg/auth/authenticator/password/ldappassword/ldap.go]: `0` is unlimitted?  Seems like we'd want to restrict that to something reasonable.

File: pkg/auth/authenticator/password/ldappassword/ldap.go [hunk 15]
```diff
@@ -0,0 +1,240 @@
...
+	glog.V(4).Infof("found dn=%q for %s", entry.DN, filter)
+
+	// Bind with given username and password to attempt to authenticate
+	if err := l.Bind(entry.DN, password); err != nil {
+		glog.V(4).Infof("error binding password for %q: %v", entry.DN, err)
+		if err, ok := err.(*ldap.Error); ok {
+			switch err.ResultCode {
+			case ldap.LDAPResultInappropriateAuthentication,
+				ldap.LDAPResultInvalidCredentials,
+				ldap.LDAPResultInsufficientAccessRights:
```
Comment (deads2k) [inline pkg/auth/authenticator/password/ldappassword/ldap.go]: `LDAPResultInsufficientAccessRights` means bad password not some sort of error?

File: pkg/auth/authenticator/password/ldappassword/ldap.go [hunk 16]
```diff
@@ -0,0 +1,240 @@
...
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/util"
+
+	"github.com/go-ldap/ldap"
+	"github.com/golang/glog"
+
+	authapi "github.com/openshift/origin/pkg/auth/api"
+	"github.com/openshift/origin/pkg/auth/authenticator"
+)
+
+const (
```
Comment (deads2k) [inline pkg/auth/authenticator/password/ldappassword/ldap.go]: duplicates from url.go?

File: pkg/auth/authenticator/password/ldappassword/ldap.go (hunk 13 above)
Comment (deads2k) [inline pkg/auth/authenticator/password/ldappassword/ldap.go]: > do we want to support arbitrary filters that allow for multiple ands/ors/nots, etc?

I see the `Filter`, but there's no option to substitute in the `username` in any location than the `QueryAttribute` predicate.

File: pkg/cmd/server/api/validation/oauth.go [hunk 17]
```diff
@@ -40,64 +41,117 @@ func ValidateOAuthConfig(config *api.OAuthConfig) fielderrors.ValidationErrorLis
...
+	u, err := ldappassword.ParseURL(provider.URL)
+	if err != nil {
+		validationResults.AddErrors(fielderrors.NewFieldInvalid("provider.url", provider.URL, err.Error()))
+		return validationResults
+	}
+
+	// Make sure bindDN and bindPassword are both set, or both unset
+	// Both unset means an anonymous bind is used for search (https://tools.ietf.org/html/rfc4513#section-5.1.1)
+	// Both set means the name/password simple bind is used for search (https://tools.ietf.org/html/rfc4513#section-5.1.3)
+	if len(provider.BindDN) > 0 && len(provider.BindPassword) == 0 {
```
Comment (deads2k) [inline pkg/cmd/server/api/validation/oauth.go]: optional: single condition check `if (len(bindDN) == 0) != (len(bindPassword) == 0) { // require both or neither`?

File: pkg/cmd/server/api/validation/oauth.go [hunk 18]
```diff
@@ -40,64 +41,117 @@ func ValidateOAuthConfig(config *api.OAuthConfig) fielderrors.ValidationErrorLis
...
+	}
+	if len(provider.BindPassword) > 0 && len(provider.BindDN) == 0 {
+		validationResults.AddErrors(fielderrors.NewFieldInvalid("provider.bindPassword", "<masked>", "Cannot specify bindPassword without bindDN"))
+	}
+
+	if provider.Insecure {
+		if u.Scheme == ldappassword.ProtocolLDAPS {
+			validationResults.AddErrors(fielderrors.NewFieldInvalid("provider.url", provider.URL, fmt.Sprintf("Cannot use %s scheme with insecure=true", u.Scheme)))
+		}
+		if len(provider.CA) > 0 {
```
Comment (deads2k) [inline pkg/cmd/server/api/validation/oauth.go]: I probably should have noticed above, but if I'm upgrading from insecure to secure, I don't get to check against a CA?

File: pkg/cmd/server/api/validation/oauth.go [hunk 19]
```diff
@@ -40,64 +41,117 @@ func ValidateOAuthConfig(config *api.OAuthConfig) fielderrors.ValidationErrorLis
...
+		}
+	} else {
+		if len(provider.CA) > 0 {
+			validationResults.AddErrors(ValidateFile(provider.CA, "provider.ca")...)
+		}
+	}
+
+	// At least one attribute to use as the user id is required
+	if len(provider.Attributes.ID) == 0 {
+		validationResults.AddErrors(fielderrors.NewFieldInvalid("provider.attributes.id", "[]", "at least one id attribute is required (LDAP standard identity attribute is 'dn')"))
```
Comment (deads2k) [inline pkg/cmd/server/api/validation/oauth.go]: required?

File: pkg/auth/authenticator/password/ldappassword/ldap.go (hunk 12 above)
Comment (liggitt) [inline pkg/auth/authenticator/password/ldappassword/ldap.go]: yes, akin to the way we validate cmd options after populating them from the command line, but don't re-validate them in their `Run()` functions when called directly from other composing commands

File: pkg/auth/authenticator/password/ldappassword/ldap.go (hunk 13 above)
Comment (liggitt) [inline pkg/auth/authenticator/password/ldappassword/ldap.go]: The existing filter can do a lot (see http://httpd.apache.org/docs/2.0/mod/mod_auth_ldap.html#examples), but is limited to attributes actually on the user record.

I don't want to bring group membership tests into the filter. `mod_auth_ldap` treats that as a separate construct. My implementation is equivalent to the "Require valid-user" directive, but they also have "Require group", etc.

Requiring a unique attribute match in the search tree on the user-specified login value doesn't seem overly burdensome... I'd leave that as-is for now.

File: pkg/auth/authenticator/password/ldappassword/ldap.go (hunk 12 above)
Comment (deads2k) [inline pkg/auth/authenticator/password/ldappassword/ldap.go]: > yes, akin to the way we validate cmd options after populating them from the command line, but don't re-validate them in their Run() functions when called directly from other composing commands

Accepted, but maybe we should fix that...

File: pkg/auth/authenticator/password/ldappassword/ldap.go (hunk 15 above)
Comment (liggitt) [inline pkg/auth/authenticator/password/ldappassword/ldap.go]: yeah, re-read the spec and reduced to `LDAPResultInvalidCredentials`

File: pkg/auth/authenticator/password/ldappassword/ldap.go (hunk 16 above)
Comment (liggitt) [inline pkg/auth/authenticator/password/ldappassword/ldap.go]: removed

File: pkg/cmd/server/api/validation/oauth.go (hunk 17 above)
Comment (liggitt) [inline pkg/cmd/server/api/validation/oauth.go]: done

File: pkg/cmd/server/api/validation/oauth.go (hunk 18 above)
Comment (liggitt) [inline pkg/cmd/server/api/validation/oauth.go]: roots are part of the tlsConfig which is used during connect or during upgrade

File: pkg/cmd/server/api/validation/oauth.go (hunk 19 above)
Comment (liggitt) [inline pkg/cmd/server/api/validation/oauth.go]: annoyingly, you don't get to provide a message with required... I can switch it if you feel strongly

File: pkg/auth/authenticator/password/ldappassword/ldap.go (hunk 14 above)
Comment (liggitt) [inline pkg/auth/authenticator/password/ldappassword/ldap.go]: documented this is left up to the LDAP server

**Comment by deads2k** [comment]
I'd really like an integration test since this seems like a likely customer configuration.  Is that proving too onerous?

**Comment by liggitt** [comment]
integration test added

**Comment by liggitt** [comment]
[test]

**Comment by deads2k** [comment]
in commit messages: `bump(package):commitID`

File: test/integration/oauth_ldap_test.go [hunk 20]
```diff
@@ -0,0 +1,180 @@
...
+		TimeLimit:    0,
+		TypesOnly:    false,
+		Attributes:   [][]byte{[]byte("cn"), []byte("dn"), []byte("mail"), []byte("myuid")},
+		Filter:       fmt.Sprintf("(&(myAttr=myValue)(myuid=%s))", myUserUID),
+	}
+
+	// Start LDAP server
+	ldapServer := testutil.NewTestLDAPServer()
+	ldapServer.SetPassword(bindDN, bindPassword)
+	ldapServer.Start("127.0.0.1:8389")
```
Comment (deads2k) [inline test/integration/oauth_ldap_test.go]: Not crazy about this.  Why not a random port?  I would like to get to a point where we can run our integration tests in parallel.

File: test/integration/oauth_ldap_test.go [hunk 21]
```diff
@@ -0,0 +1,180 @@
...
+	testutil "github.com/openshift/origin/test/util"
+
+	"github.com/vjeantet/ldapserver"
+)
+
+func TestOAuthLDAP(t *testing.T) {
+	const (
+		providerName = "myldapprovider"
+
+		bindDN       = "o=company,uid=admin"
```
Comment (deads2k) [inline test/integration/oauth_ldap_test.go]: Add some sort of randomization to this and the password.  I had a bad experience and now I'm afraid of hardcoded usernames and passwords like this in tests.

File: test/integration/oauth_ldap_test.go [hunk 22]
```diff
@@ -0,0 +1,180 @@
...
+		bindPassword = "admin-password"
+
+		myUserUID      = "myuser"
+		myUserName     = "My User, Jr."
+		myUserEmail    = "myuser@example.com"
+		myUserDN       = "o=company,myuid=" + myUserUID
+		myUserPassword = "myuser-password"
```
Comment (deads2k) [inline test/integration/oauth_ldap_test.go]: I don't care about this one, just the ones that go in our config.

File: test/integration/oauth_ldap_test.go (hunk 21 above)
Comment (liggitt) [inline test/integration/oauth_ldap_test.go]: added a random suffix

File: test/integration/oauth_ldap_test.go (hunk 22 above)
Comment (liggitt) [inline test/integration/oauth_ldap_test.go]: I went ahead and randomized them all

File: test/integration/oauth_ldap_test.go (hunk 20 above)
Comment (liggitt) [inline test/integration/oauth_ldap_test.go]: fixed

**Comment by deads2k** [comment]
lgtm [merge]

**Comment by liggitt** [comment]
```
error: error when creating "examples/sample-app/application-template-dockerbuild.json": Post https://127.0.0.1:8443/osapi/v1beta3/namespaces/default/templates: net/http: TLS handshake timeout
!!! Error in hack/test-cmd.sh:240
  'oc create -f examples/sample-app/application-template-dockerbuild.json' exited with status 1
Call stack:
  1: hack/test-cmd.sh:240 main(...)
```

**Comment by liggitt** [comment]
removing merge tag, completing testing of BER-decoding error recovery in added godep

**Comment by liggitt** [comment]
[merge]
